client/finality-grandpa/src/communication/gossip.rs

```diff
@@ -539,11 +540,20 @@ impl<N: Ord> Peers<N> {
 			Some(p) => p,
 		};
 
-		let invalid_change = peer.view.set_id > update.set_id ||
-			peer.view.round > update.round && peer.view.set_id == update.set_id ||
-			peer.view.last_commit.as_ref() > Some(&update.commit_finalized_height);
+		let set_id_increased = update.set_id > peer.view.set_id;
+		let round_in_set_increased =
+			update.set_id == peer.view.set_id && update.round > peer.view.round;
+		let finalized_height_in_round_increased = update.set_id == peer.view.set_id &&
+			update.round == peer.view.round &&
+			Some(&update.commit_finalized_height) > peer.view.last_commit.as_ref();
+		let finalized_height_not_decreased =
+			Some(&update.commit_finalized_height) >= peer.view.last_commit.as_ref();
 
-		if invalid_change {
+		let valid_change = (set_id_increased || round_in_set_increased) &&
+			finalized_height_not_decreased ||
+			finalized_height_in_round_increased;
+
+		if !valid_change {
 			return Err(Misbehavior::InvalidViewChange)
```

Comment:
I think it might make sense to create a new misbehavior variant for this, maybe `DuplicateNeighborMessage`, which would have a lower punishment cost (maybe -100?). This is because we have logic to periodically rebroadcast neighbor messages in case there is no progress happening (https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/communication/periodic.rs#L57-L60), so there might be situations where nodes legitimately send out a duplicate neighbor message and I think we want to reduce the reputation less harsh then when an actual invalid view change happens. This would still address the problem of someone spamming neighbor messages.

Reply:
Should we align the period of rebroadcasting neighbor messages with punishment logic then? Like, punish only peers that spam us more frequently than the rebroadcast period?

Reply:
That would certainly be better than what I suggested! Keeping track of when we got the last neighbor message per peer and then if we receive a duplicate after `PERIODIC_REBROADCAST` it is valid, otherwise it's spam.

Reply:
Done.